### ISSUE TITLE
Crevice top left and right corners and top half

### ISSUE THREAD

**Issue by rogierstuyver** [eczarny/spectacle#321]
When aligning a window to the top left or right corner or to the top half the window does not reach the menubar completely, leaving a small crevice. See the attached screenshot where the window on the left is aligned top left and the window on the right is aligned right half, correctly reaching the menubar. 

This is using Spectacle 0.8.8 and Yosemite 10.10.2 and happens on all applications I used (including MS Word, Safari, Preview, TeXShop and Finder). 

![screen shot 2015-02-09 at 16 50 32](https://cloud.githubusercontent.com/assets/10925628/6109642/959c1d0c-b07c-11e4-9bc6-4a9bda68a538.png)

Might be related to #243.

**Comment by janniklorenz** [comment]
I found that in the SpectacleWindowPositionManager the lines 243 to 278 (https://github.com/eczarny/spectacle/blob/master/Spectacle/SpectacleWindowPositionManager.m#L243), which Provide the 1/3 and 2/3 functionality are the source of the Problem. Uncomment fixed it for me.

```
if ((action != SpectacleWindowActionUndo) && (action != SpectacleWindowActionRedo)) {
        [self moveWindowRect:windowRect frontMostWindowElement:frontMostWindowElement];

        movedWindowRect = [self rectOfWindowWithAccessibilityElement: frontMostWindowElement];
        
        // Does the window fit within the desired position?
        if (!CGRectEqualToRect(movedWindowRect, windowRect)) {
            CGRect adjustedWindowRect = windowRect;
            
            // If not, try reducing the window size to fit.
            while (movedWindowRect.size.width > windowRect.size.width || movedWindowRect.size.height > windowRect.size.height) {
                if (movedWindowRect.size.width > windowRect.size.width) {
                    adjustedWindowRect.size.width -= 2;
                }

                if (movedWindowRect.size.height > windowRect.size.height) {
                    adjustedWindowRect.size.height -= 2;
                }
                
                // If the window's size has been reduced to half of its original size, stop.
                if (adjustedWindowRect.size.width < windowRect.size.width / 2.0f || adjustedWindowRect.size.height < windowRect.size.height / 2.0f) {
                    break;
                }
                
                [self moveWindowRect:adjustedWindowRect frontMostWindowElement:frontMostWindowElement];

                movedWindowRect = [self rectOfWindowWithAccessibilityElement: frontMostWindowElement];
            }
            
            // Center the window, taking into account any quantization adjustments.
            adjustedWindowRect.origin.x += floor((windowRect.size.width - movedWindowRect.size.width) / 2.0f);
            adjustedWindowRect.origin.y += floor((windowRect.size.height - movedWindowRect.size.height) / 2.0f);
            
            [self moveWindowRect:adjustedWindowRect frontMostWindowElement:frontMostWindowElement];
        }
    }
```

**Comment by DerrickGold** [comment]
Hey janniklorenz, I tried messing around with the code in that area to see if it affected the issue at all and had no luck. However, I have done some further investigating and believe my original hunch of an off by one error causing the issue may still hold true, but a different fix from my original would be applied.

In SpectacleWindowPositionCalculator.m, the origin for top half windows is calculated using:
```objective-c
if (MovingToTopRegionOfDisplay(action)) {
        calculatedWindowRect.origin.y = visibleFrameOfScreen.origin.y + floor((visibleFrameOfScreen.size.height) / 2.0f);
}
```
Given that the screen height falls in a range of (0 to MaxHeight - 1), and the menu bar having an even numbered height, the calculation cannot divide the visibleFrameOfScreen.size.height evenly. The result is rounded down making one half of the screen larger than the other.

For example, with my screen of resolution 1920x1200, the Y resolution ranges from (0 to 1199), and the visibleFrameOfScreen Y resolution range calculates to (0 - 1177). Half of the visible space, 1177 / 2 = 588.5, then gets rounded down to 588, so the origin and window height are now calculated. But, when you calculate the new origin for the top half of the screen, it is bottom + 588, the top half of the screen is the remaining 589 pixels since we started with an odd resolution number. 

Simply adding one to the resolution height while doing the calculation like this:
```objective-c
if (MovingToTopRegionOfDisplay(action)) {
        calculatedWindowRect.origin.y = visibleFrameOfScreen.origin.y + floor((visibleFrameOfScreen.size.height + 1) / 2.0f);
}
```
will solve the problem, but I'm not quite sure if this is the exact fix would be preferred as one could also fix the issue by bumping the window height by one for this case, or like in my original fix, just bump up the resolution to something even for all calculations.

**Comment by eczarny** [comment]
I think I have a fix on master. Could folks try it out?

**Comment by janniklorenz** [comment]
That works pretty good for me. 

I found just one position, which isn't 100% accurate. When I resize Xcode to 1/3 down with a visible Dock, the window is a litte bit to deep. That only happens with Xcode, the Finder for example works well.

Has someone the same issue?

**Comment by eczarny** [comment]
Oh weird, I'm having major issues with Xcode. Looks like there may be a regression someplace. I'll investigate.

**Comment by eczarny** [comment]
Okay, I think I have a fix in place. I reverted 5517a9644de282c722f767dbcc79a09e86677bc0 as it was introducing less than ideal behavior. The functionality that commit implemented will be tracked through a separate issue and evaluated for a future release.

Could you try out Spectacle from master with Xcode and a visible Dock again?

**Comment by janniklorenz** [comment]
Sure. The Xcode window still is to deep, its nearly 80% under the dock. Did this commit worked for you?
![bildschirmfoto 2015-04-16 um 21 34 12](https://cloud.githubusercontent.com/assets/1284753/7189716/0bf40fb8-e481-11e4-847d-7836a3070a44.png)

**Comment by eczarny** [comment]
Ah, okay. I fixed another issue but unrelated to this particular problem. Thanks for including the screenshot! It helps me narrow down the incorrect behavior.

**Comment by eczarny** [comment]
Spectacle should not allow you to resize Xcode to vertical thirds. This functionality was added when I merged in #309 so I'm going to create a separate issue to track that particular problem. I'm closing this issue since I believe it is fixed.
